src/content/docs/en/reference/integrations-reference.mdx

```diff
@@ -613,6 +613,121 @@ A list of all generated pages. It is an object with one property.
+
+#### Defining hook types
+
+Provide the type for your custom hook by augmenting Astro’s `IntegrationHooks` interface within your integration:
+
+```ts title="types.ts"
+// on an implementation file
+declare global {
+  namespace Astro {
+    interface IntegrationHooks {
+      'myLib:eventHappened': (your: string, parameters: number) => Promise<void>;
+    }
+  }
+}
+```
+
+#### Calling custom hooks
+
+Installed integrations can be accessed from the `astro:config:setup` and `astro:config:done` hooks. Loop over these to call your custom hook on each if available:
+
+```ts {7-9}
+// your-integration/index.ts
+export default (): AstroIntegration => {
+  return {
+    name: 'your-integration',
+    hooks: {
+      'astro:config:setup': async ({ config }) => {
+        for (const integration of config.integrations) {
+          await integration.hooks['myLib:eventHappened'].?('your values', 123);
+        }
+      },
+    }
+  }
+}
+```
+
+To call custom hooks outside of `astro:config:setup` and `astro:config:done`, store the integrations array and use it later. `astro:config:done` should be preferred to work with integrations that get added by other integrations during `astro:config:setup`.
+
+```ts {3,9-10,13-16}
+// your-integration/index.ts
+export default (): AstroIntegration => {
+  let integrations: AstroIntegration[] = [];
+
+  return {
+    name: 'your-integration',
+    hooks: {
+      'astro:config:done': async ({ config }) => {
+        // Store the integrations when available
+        integrations = config.integrations;
+      },
+      'astro:build:done': async () => {
+        // Loop over the stored integrations in a later hook
+        for (const integration of config.integrations) {
+          await integration.hooks['myLib:eventHappened'].?('your values', 123);
+        }
+      },
+    }
+  }
+}
+```
+
+#### Using a custom hook from another integration
+
+An integration that uses your custom hook will need to import your types file. It can then provide a value for your hook in `hooks` with type checking and editor hints.
+
+```ts {6-8}
+// other-integration/index.ts
+export default (): AstroIntegration => {
+  return {
+    name: 'other-integration',
+    hooks: {
+      'myLib:eventHappened': async (your, values) => {
+        // ...
+      },
+    }
+  }
+}
+```
+
+Make sure that your type augmentation can be loaded by other integrations; otherwise, their hook implementations won't be properly typed. The easiest way to ensure this is to either have that declaration in the same file that exports your integration or import it from that file:
+
+```ts {2-3} {5-12}
+// your-integration/index.ts
+// Import the types
+import './types.ts';
+
+// Or declare it in the same file
```

Comment:
As Chris mentioned on Discord, we would prefer not to have two conflicting bits of code in the same code sample. You have the power to just choose one method to show! Alternatively, you can present the alternative separately. (But it's honestly fine to just show one thing if it will always work)